speechbrain/utils/autocast.py

```diff
@@ -3,13 +3,131 @@
 
 import torch
+from contextlib import nullcontext
 
+@dataclass
+class AMPConfig:
+    """Configuration for automatic mixed precision (AMP).
+
+    Arguments
+    ---------
+    dtype : torch.dtype
+        The dtype to use for AMP.
+    """
+
+    dtype: torch.dtype
+
+    @classmethod
+    def from_name(self, name):
+        """Create an AMPConfig from a string name.
+
+        Arguments
+        ---------
+        name : str
+            The name of the AMPConfig to create.  Must be one of `fp32`,
+            `fp16`, or `bf16`.
+
+        Returns
+        -------
+        AMPConfig
+            The AMPConfig corresponding to the name.
+        """
+        if name is None or name == "fp32":
+            return AMPConfig(torch.float32)
+        elif name == "fp16":
+            return AMPConfig(torch.float16)
+        elif name == "bf16":
+            return AMPConfig(torch.bfloat16)
+        else:
+            raise ValueError(
+                f"Specified autocast mode ({name}) incorrect, expected one of `fp32`, `fp16`, `bf16`."
+            )
+        
+class TorchAutocast:
+    """
+    Utility for conditionally enabling torch.autocast.
+
+    This context manager activates autocasting (via ``torch.autocast``)
+    when the ``enabled`` flag is True. When disabled, it behaves as a no-op
+    context manager.
+
+    Arguments
+    ---------
+    enabled : bool
+        Whether to enable torch.autocast.
+    *args
+        Positional arguments to pass to torch.autocast.
+    **kwargs
+        Keyword arguments to pass to torch.autocast.
+
+    Examples
+    --------
+    >>> with TorchAutocast(enabled=True, device="cpu", dtype=torch.bfloat16):
+    ...     # Run model inference with autocasting enabled.
+    """
+
+    def __init__(self, enabled: bool, *args, **kwargs):
+        if enabled:
+            self.context = torch.autocast(*args, **kwargs)
+        else:
+            self.context = nullcontext()  # no-op context manager
+
+    def __enter__(self):
+        """
+        Enter the autocast context.
+
+        Returns
+        -------
+        context
+            The result of entering the underlying autocast context manager.
+
+        Raises
+        ------
+        RuntimeError
+            If an error occurs while entering the autocast context and the context
+            provides 'device' and 'fast_dtype' attributes, a RuntimeError is raised
+            with additional diagnostic information.
+        """
+        try:
+            return self.context.__enter__()
+        except RuntimeError as e:
+            if hasattr(self.context, "device") and hasattr(self.context, "fast_dtype"):
+                device = self.context.device
+                dtype = self.context.fast_dtype
+                raise RuntimeError(
+                    f"Error during autocasting with dtype={dtype} on device={device}.\n"
+                    "If you are on the FAIR Cluster, you might need to use autocast_dtype=float16."
```

Comment:
Do we really want a cluster-specific message here? Perhaps we could make it a bit more general, like "Ensure your hardware supports the dtype you are attempting to use."

Reply:
oops this is a reliquat from some code i have stolen / adapted